### ISSUE TITLE
Mismatch between local Nominatim installation and OpenStreetMap

### ISSUE THREAD

**Issue by cjgibson** [openstreetmap/Nominatim#692]
Running the query [`q=广东广州`](https://nominatim.openstreetmap.org/search.php?q=%E5%B9%BF%E4%B8%9C%E5%B9%BF%E5%B7%9E&format=json) against OpenStreetMap returns ["Guangzhou City, Guangdong, China"](https://nominatim.openstreetmap.org/details.php?osmtype=N&osmid=244080443), which is exactly what I expected. However, running this same query against a freshly built Nominatim instance that I have locally returns ["昼"](https://nominatim.openstreetmap.org/details.php?osmtype=N&osmid=3112130629), and nothing else.

Looking into the debugging output from both my local machine and OpenStreetMap's API, I believe I have found the mismatch causing my odd local result. The initial PostgreSQL query for the tokens within the search itself is as follows:

```
SELECT word_id, word_token, word, class, type, country_code, operator, search_name_count
FROM word
WHERE word_token in (' yan dong yan zhou','yan dong yan zhou',' yan','yan',' dong yan zhou',
                     'dong yan zhou',' dong','dong',' yan zhou','yan zhou',' zhou','zhou',
                     ' dong yan','dong yan',' yan dong','yan dong',' yan dong yan','yan dong yan')
```

Which, when run against my local installation, returns:
```
 word_id  | word_token | word  | class | type | country_code | operator | search_name_count 
----------+------------+-------+-------+------+--------------+----------+-------------------
      265 | dong       |       |       |      |              |          |             69841
  1625808 |  dong      | 咚喇  |       |      |              |          |                 0
 16451860 |  dong yan  | 董言  |       |      |              |          |                 0
 16451861 | dong yan   |       |       |      |              |          |                 0
      231 | yan        |       |       |      |              |          |             21926
   561608 |  yan       | Ya'en |       |      |              |          |                 0
  1434234 |  yan dong  | 烟洞  |       |      |              |          |                 0
  1434235 | yan dong   |       |       |      |              |          |                 0
 16335319 |  yan zhou  | 广州  |       |      |              |          |                 0
  3610417 | yan zhou   |       |       |      |              |          |                 0
    10368 | zhou       |       |       |      |              |          |             13275
  1849783 |  zhou      | 宙    |       |      |              |          |                 0
(12 rows)
```

From what I can tell, the same query against the PostgreSQL instance utilized by OpenStreetMap's API returns something like the following:

```
 word_id  | word_token | word  | class | type | country_code | operator | search_name_count 
----------+------------+-------+-------+------+--------------+----------+-------------------
      260 | dong       |       |       |      |              |          |             69841
  1535579 |  dong      | 咚喇  |       |      |              |          |                 0
 17045649 |  dong yan  | 董言  |       |      |              |          |                 0
  6029436 | dong yan   |       |       |      |              |          |                 0
      229 | yan        |       |       |      |              |          |             21926
   530235 |  yan       | Ya'en |       |      |              |          |                 0
  8343004 |  yan dong  | 广洞  |       |      |              |          |                 0
  7208058 | yan dong   |       |       |      |              |          |                 0
  7859650 |  yan zhou  | 广州市 |       |      |              |          |                 0
 16936362 | yan zhou   |       |       |      |              |          |                 0
     9803 | zhou       |       |       |      |              |          |             13275
  1747446 |  zhou      | 宙    |       |      |              |          |                 0
(12 rows)
```

With the important differences being the rows:
```
 word_token | local_id | local_word |   OSM_id | OSM_word
------------+----------+------------+----------+----------
  yan dong  |  1434234 | 烟洞       |  8343004 | 广洞
  yan zhou  | 16335319 | 广州       |  7859650 | 广州市
```

As a result of this mismatch, the final search query processed while attempting to geocode the input (presented in the [debugging output as "Search Loop, group 3, loop 21"](https://nominatim.openstreetmap.org/search.php?q=广东广州&debug=1)) returns an array of six place IDs via the OpenStreetMap API and none locally.

Long story short, just wondering how the `word` table is sourced, and if there are any means by which I can correct the corrupted rows present within my local installation. As far as I can tell, this issue mostly affects queries containing Chinese input, but am admittedly unable to verify results in languages other than English and Chinese.

**Comment by lonvia** [comment]
there is a fundamental problem with matching Chinese characters here. 广东 does not appear as a word in the OSM data, there is only '广东省', so to find it Nominatim has to resort to partial matches, i.e. finding matches that have '广' and '东' in their name. Unfortunately we do that in the transliterated form 'yan' and 'dong' which both are words that appear so often that they are considered stop words and do not appear in the search index, so ' 广东' can never be found. What comes back as a result is essentially random and may vary from installation to installation (it depends on the order by which words are retrieved from the word table).

Nominatim would actually have to do the search over the original characters to make partial search work in Chinese. I'm not sure how that can work with the current implementation.

Quick fix: add a word variant to the provinces in OSM, e.g. short_name='广东'.

**Comment by cjgibson** [comment]
Ah, understood. And due to the nature of Chinese text, partial matches decay into a series of lone-character match logic, makes sense. If you have an entry-point inside Nominatim's query logic for language-specific token terminals, `省` would make for a solid addition. That character literally means "province" in Chinese, so `广东省`, becomes `广东 省`, interpreted as "Guangdong Province". You also see this occur with cities: `市` literally translates as city, and you'll see `武汉市`, or `武汉 市`, interpreted as "Wuhan City".

Somewhat obnoxiously, you'll also see both location types used without the trailing identifier. `广东广州`, for instance, is best interpreted as `广东省, 广州市`, or "Guangdong Province, Guangzhou City". (If I'm allowed a self-reference, this kind of tokenization and type-tagging-by-terminal also solves issue #658).

A proper solution would likely be a series of language-specific tokenization engines (at least for CJK), applied and generated specifically for tokenizing query strings; but that may lie beyond the scope of this project.

**Comment by lonvia** [comment]
The tokenization happens in [the module code](https://github.com/openstreetmap/Nominatim/tree/master/module) where the Chinese characters are romanized with an additional space at the end to make them later count as single words.

We already have an exception in place for `市` [here](https://github.com/openstreetmap/Nominatim/blob/master/sql/functions.sql#L320). I'm a little bit reluctant to add further exceptions. As you say, this needs a more general solution. The problem is actually not so different from the one where an object is tagged with say 'best western hotel' and you want to be able to find it as 'best western'. It just so happens that partial matching works a little bit better with Latin characters.
